packages/firestore/test/unit/util/misc.test.ts

```diff
@@ -53,3 +53,216 @@ describe('FieldMask', () => {
+    } else {
+      this.rnd = seedOrRnd;
+      this.surrogatePairProbability = StringGenerator.validateProbability(
+        surrogatePairProbability!
+      );
+      this.maxLength = StringGenerator.validateLength(maxLength!);
+    }
+  }
+
+  private static validateProbability(probability: number): number {
+    if (!Number.isFinite(probability)) {
+      throw new Error(
+        `invalid surrogate pair probability: ${probability} (must be between 0.0 and 1.0, inclusive)`
+      );
+    } else if (probability < 0.0) {
+      throw new Error(
+        `invalid surrogate pair probability: ${probability} (must be greater than or equal to zero)`
+      );
+    } else if (probability > 1.0) {
+      throw new Error(
+        `invalid surrogate pair probability: ${probability} (must be less than or equal to 1)`
+      );
+    }
+    return probability;
+  }
+
+  private static validateLength(length: number): number {
+    if (length < 0) {
+      throw new Error(
+        `invalid maximum string length: ${length} (must be greater than or equal to zero)`
+      );
+    }
+    return length;
+  }
+
+  next(): string {
+    const length = this.rnd.nextInt(this.maxLength + 1);
+    const sb = new StringBuilder();
+    while (sb.length() < length) {
+      const codePoint = this.nextCodePoint();
+      sb.appendCodePoint(codePoint);
+    }
+    return sb.toString();
+  }
+
+  private isNextSurrogatePair(): boolean {
+    return StringGenerator.nextBoolean(this.rnd, this.surrogatePairProbability);
+  }
+
+  private static nextBoolean(rnd: Random, probability: number): boolean {
+    if (probability === 0.0) {
+      return false;
+    } else if (probability === 1.0) {
+      return true;
+    } else {
+      return rnd.nextFloat() < probability;
+    }
+  }
+
+  private nextCodePoint(): number {
+    if (this.isNextSurrogatePair()) {
+      return this.nextSurrogateCodePoint();
+    } else {
+      return this.nextNonSurrogateCodePoint();
+    }
+  }
+
+  private nextSurrogateCodePoint(): number {
+    const highSurrogateMin = 0xd800;
+    const highSurrogateMax = 0xdbff;
+    const lowSurrogateMin = 0xdc00;
+    const lowSurrogateMax = 0xdfff;
+
+    const highSurrogate = this.nextCodePointRange(
+      highSurrogateMin,
+      highSurrogateMax
+    );
+    const lowSurrogate = this.nextCodePointRange(
+      lowSurrogateMin,
+      lowSurrogateMax
+    );
+
+    return (highSurrogate - 0xd800) * 0x400 + (lowSurrogate - 0xdc00) + 0x10000;
+  }
+
+  private nextNonSurrogateCodePoint(): number {
+    let codePoint;
+    do {
+      codePoint = this.nextCodePointRange(0, 0xffff); // BMP range
+    } while (codePoint >= 0xd800 && codePoint <= 0xdfff); // Exclude surrogate range
+
+    return codePoint;
+  }
+
+  private nextCodePointRange(min: number, max: number): number {
+    const rangeSize = max - min + 1;
+    const offset = this.rnd.nextInt(rangeSize);
+    return min + offset;
+  }
+}
+
+class Random {
+  private seed: number;
```

Comment:
don't need a seed.

Reply:
seed id used to track the exact test case that is failing.